File: solutions/ruby/grep/1/grep.rb
# frozen_string_literal: true

# module Grep
module Grep
  Match = Struct.new(:filename, :index, :line)

  def self.filegrep(flags, pattern, filename)
    File.open(filename).each_line(chomp: true).zip(1..).to_enum.filter_map do |line, index|
      match_line = line
      match_line = match_line.downcase if flags.include?('-i')
      Match.new(filename, index, line) if flags.include?('-v') == !match_line.match(pattern)
    end
  end

  def self.build_pattern(pattern, flags)
    pattern = "^#{pattern}$" if flags.include?('-x')
    pattern = pattern.downcase if flags.include?('-i')
    pattern
  end

  def self.grep(pattern, flags, files)
    pattern = build_pattern(pattern, flags)
    files.flat_map { filegrep(flags, pattern, _1) }.map do |match|
      line = match.line
      line = "#{match.index}:#{line}" if flags.include?('-n')
      line = "#{match.filename}:#{line}" if files.count > 1
      line = match.filename if flags.include?('-l')
      line
    end.uniq.join("\n")
  end
end

File: solutions/lua/grep/1/grep.lua
function table.contains(tbl, element)
	for _, value in pairs(tbl) do
		if value == element then return true end
	end
	return false
end

local function grep(result, file, pattern, flags)
	local n = 1
	for line in io.lines(file) do
		local match_line = line
		if flags:contains("-i") then match_line = string.lower(line) end
		if flags:contains("-v") then
			if not match_line:match(pattern) then result:insert({ file, n, line }) end
		else
			if match_line:match(pattern) then result:insert({ file, n, line }) end
		end
		n = n + 1
	end
end

local function result(matches, files, flags)
	local r = setmetatable({}, { __index = table })
	for i = 1, #matches do
		local line = matches[i][3]
		if flags:contains("-n") then line = matches[i][2] .. ":" .. line end
		if #files > 1 then line = matches[i][1] .. ":" .. line end
		if flags:contains("-l") then line = matches[i][1] end
		if not r:contains(line) then r:insert(line) end
	end
	return r
end

return function(options)
	local flags = setmetatable(options.flags, { __index = table })
	-- parse the flags
	if flags:contains("-x") then options.pattern = "^" .. options.pattern .. "$" end
	if flags:contains("-i") then options.pattern = string.lower(options.pattern) end
	-- do the grepping
	local matches = setmetatable({}, { __index = table })
	for _, file in ipairs(options.files) do
		grep(matches, file, options.pattern, flags)
	end
	-- prepare results
	return result(matches, options.files, flags)
end
